File: src/game/achievements/progress.ts
import { getSkillProgress } from '../activity/xp'
import { CRITTER_DEFS, collectionCount } from '../critters/critters'
import type { GameDatabase } from '../data/types'
import { itemHasCapability, WEAPON_TOOL_SLOT_ID } from '../equipment/loadout'
import { totalLevel, totalSkillXp } from '../skills/totals'
import type { AchievementProgress, PlayerSave } from '../save/types'

export interface AchievementRow {
  'Achievement ID': string
  'Internal Key': string
  'Display Name': string
  Category: string | null
  Difficulty: string | null
  'Check Type': string | null
  'Target ID': string | null
  'Required Count': number | null
  Status: string
  'Release Phase': string
  Reward: string | null
  'Target Skill ID': string | null
  'Required Level': number | null
  Notes: string | null
}

export const ACHIEVEMENT_DIFFICULTIES = ['Easy', 'Medium', 'Hard'] as const

export function addLifetimeStat(save: PlayerSave, key: string, amount = 1): PlayerSave {
  const current = Number(save.statistics.values[key] ?? 0)
  return {
    ...save,
    statistics: {
      values: {
        ...save.statistics.values,
        [key]: current + amount,
      },
    },
  }
}

export function isSpellProject(project: { 'Internal Key'?: string; 'Display Name'?: string }): boolean {
  const key = String(project['Internal Key'] ?? '')
  const name = String(project['Display Name'] ?? '').toLowerCase()
  return key.includes('_spell') || (name.includes('spell') && !name.includes('enchant'))
}

export function recordProjectMilestones(
  db: GameDatabase,
  save: PlayerSave,
  projectId: string,
  crafts: number,
): PlayerSave {
  const project = db.Projects.find((row) => row['Project ID'] === projectId)
  let next = addLifetimeStat(save, `project_${projectId}`, crafts)
  const locationId = save.currentLocationId
  if (locationId) {
    next = addLifetimeStat(next, `project_${projectId}_at_${locationId}`, crafts)
  }
  const outputId = String(project?.['Output Item / Target ID'] ?? '')
  if (outputId.startsWith('ENCH-')) {
    next = addLifetimeStat(next, 'items_enchanted', crafts)
  }
  if (project && isSpellProject(project)) {
    next = addLifetimeStat(next, 'spell_projects', crafts)
  }
  return next
}

export function recordProductionMilestones(
  db: GameDatabase,
  save: PlayerSave,
  outputItemId: string,
  quantity: number,
): PlayerSave {
  let next = addLifetimeStat(save, `output_${outputItemId}`, quantity)
  const locationId = save.currentLocationId
  if (locationId) {
    next = addLifetimeStat(next, `output_${outputItemId}_at_${locationId}`, quantity)
  }
  const item = db.Items.find((row) => row['Item ID'] === outputItemId)
  if (item?.Category === 'Potion') {
    next = addLifetimeStat(next, 'potions_created', quantity)
  }
  return next
}

export function recordFoodConsumed(save: PlayerSave, itemId: string): PlayerSave {
  return addLifetimeStat(save, `consumed_${itemId}`, 1)
}

export function recordGatheredDrops(
  save: PlayerSave,
  itemIds: Iterable<string>,
  locationId: string,
  weaponId: string | null,
): PlayerSave {
  const wield = weaponId && weaponId.length > 0 ? weaponId : 'none'
  let next = save
  for (const itemId of itemIds) {
    next = addLifetimeStat(next, `gathered_${itemId}_at_${locationId}_wield_${wield}`)
  }
  return next
}

export function recordItemsSoldAtLocation(
  save: PlayerSave,
  items: Iterable<{ itemId: string; quantity: number }>,
  locationId: string,
): PlayerSave {
  let next = save
  for (const item of items) {
    const qty = Math.max(0, Number(item.quantity) || 0)
    if (qty <= 0) continue
    next = addLifetimeStat(next, `sold_${item.itemId}_at_${locationId}`, qty)
  }
  return next
}

function splitTags(value: string | null | undefined): string[] {
  if (typeof value !== 'string') return []
  return value
    .split(';')
    .map((part) => part.trim().toLowerCase())
    .filter(Boolean)
}

/** True when the item carries a class label such as `staff` or `wand`. */
export function itemHasClassLabel(db: GameDatabase, itemId: string, label: string): boolean {
  const wanted = label.trim().toLowerCase()
  if (!wanted) return false
  if (itemHasCapability(db, itemId, wanted)) return true
  const item = db.Items.find((row) => row['Item ID'] === itemId)
  return splitTags(item?.['Functional / Source Tags']).includes(wanted)
}

export function classLabelsFromAchievements(db: GameDatabase): string[] {
  const labels = new Set<string>()
  for (const row of asAchievementRows(db)) {
    if (row['Check Type'] !== 'kill_enemy_class') continue
    const parsed = parseKillEnemyClass(row['Target ID'])
    parsed?.classes.forEach((name) => labels.add(name))
  }
  return [...labels]
}

export function classLabelsOnItem(db: GameDatabase, itemId: string): string[] {
  return classLabelsFromAchievements(db).filter((label) => itemHasClassLabel(db, itemId, label))
}

export function recordEnemyKill(db: GameDatabase, save: PlayerSave, enemyId: string): PlayerSave {
  let next = addLifetimeStat(save, `killed_${enemyId}`)
  const weaponId = save.equipment.slots[WEAPON_TOOL_SLOT_ID]?.itemId
  if (!weaponId) return next
  for (const label of classLabelsOnItem(db, weaponId)) {
    next = addLifetimeStat(next, `killed_${enemyId}_class_${label}`)
  }
  return next
}

function parseAtLocation(target: string | null): { id: string; locationId: string } | null {
  if (!target) return null
  const at = target.indexOf('@')
  if (at <= 0 || at === target.length - 1) return null
  return { id: target.slice(0, at), locationId: target.slice(at + 1) }
}

function parseGatherDrop(
  target: string | null,
): { itemId: string; locationId: string; weaponId: string } | null {
  if (!target) return null
  const [atPart, wieldPart] = target.split('+wield:')
  const parsed = parseAtLocation(atPart ?? null)
  if (!parsed || !wieldPart) return null
  return { itemId: parsed.id, locationId: parsed.locationId, weaponId: wieldPart }
}

function parseKillEnemyClass(
  target: string | null,
): { enemyId: string; classes: string[] } | null {
  if (!target) return null
  const [enemyId, rest] = target.split('+class:')
  if (!enemyId || !rest) return null
  const classes = rest
    .split('|')
    .map((part) => part.trim().toLowerCase())
    .filter(Boolean)
  if (classes.length === 0) return null
  return { enemyId, classes }
}

function equippedItemIds(save: PlayerSave): Set<string> {
  const ids = new Set<string>()
  for (const stack of Object.values(save.equipment.slots)) {
    if (stack?.itemId) ids.add(stack.itemId)
  }
  return ids
}

function holdsEquipSet(save: PlayerSave, target: string | null): boolean {
  if (!target) return false
  const needed = target
    .split(',')
    .map((part) => part.trim())
    .filter(Boolean)
  if (needed.length === 0) return false
  const worn = equippedItemIds(save)
  return needed.every((itemId) => worn.has(itemId))
}

export interface StatisticRow {
  'Statistic ID': string
  'Internal Key': string
  'Display Name': string
  Category: string | null
  Status: string
  'Release Phase': string
  Notes: string | null
}

export function asAchievementRows(db: GameDatabase): AchievementRow[] {
  return db.Achievements as unknown as AchievementRow[]
}

export function asStatisticRows(db: GameDatabase): StatisticRow[] {
  return db.Statistics as unknown as StatisticRow[]
}

/**
 * Achievements in this category are re-checked every sync and can be lost.
 *
 * A skill milestone is a thing the player did once, so it is theirs forever. A
 * collection is a statement about the collection as it stands now, which stops
 * being true the moment the world grows a new critter.
 */
export const REVOCABLE_ACHIEVEMENT_CATEGORY = 'Collections'

export const CRITTER_COLLECTOR_ACHIEVEMENT_ID = 'ACH-0015'

function upsertAchievement(
  list: AchievementProgress[],
  achievementId: string,
  unlockedAt: string,
): AchievementProgress[] {
  const existing = list.find((row) => row.achievementId === achievementId)
  if (existing?.unlocked) return list
  const next = list.filter((row) => row.achievementId !== achievementId)
  next.push({ achievementId, unlocked: true, unlockedAt })
  return next
}

function revokeAchievement(
  list: AchievementProgress[],
  achievementId: string,
): AchievementProgress[] {
  if (!list.some((row) => row.achievementId === achievementId)) return list
  return list.filter((row) => row.achievementId !== achievementId)
}

/** Whether the collection holds at least one of every critter that exists. */
export function hasEveryCritter(save: PlayerSave): boolean {
  if (CRITTER_DEFS.length === 0) return false
  return CRITTER_DEFS.every((critter) => collectionCount(save, critter.id) > 0)
}

/** Whether a save currently qualifies for a category that can be lost again. */
function holdsRevocableAchievement(save: PlayerSave, achievementId: string): boolean {
  return achievementId === CRITTER_COLLECTOR_ACHIEVEMENT_ID && hasEveryCritter(save)
}

function lifetimeCount(values: Record<string, number>, key: string): number {
  return Number(values[key] ?? 0)
}

function holdsMilestone(
  db: GameDatabase,
  save: PlayerSave,
  achievement: AchievementRow,
  values: Record<string, number>,
): boolean {
  const check = achievement['Check Type'] ?? ''
  const target = achievement['Target ID']
  const count = Number(achievement['Required Count'] ?? 1)
  const required = achievement['Required Level']
  switch (check) {
    case 'project':
      return target != null && lifetimeCount(values, `project_${target}`) >= count
    case 'consume':
      return target != null && lifetimeCount(values, `consumed_${target}`) >= count
    case 'output_item':
      return target != null && lifetimeCount(values, `output_${target}`) >= count
    case 'output_at_location': {
      const parsed = parseAtLocation(target)
      return parsed != null && lifetimeCount(values, `output_${parsed.id}_at_${parsed.locationId}`) >= count
    }
    case 'project_at_location': {
      const parsed = parseAtLocation(target)
      return (
        parsed != null && lifetimeCount(values, `project_${parsed.id}_at_${parsed.locationId}`) >= count
      )
    }
    case 'gather_drop': {
      const parsed = parseGatherDrop(target)
      return (
        parsed != null &&
        lifetimeCount(
          values,
          `gathered_${parsed.itemId}_at_${parsed.locationId}_wield_${parsed.weaponId}`,
        ) >= count
      )
    }
    case 'sold_at_location': {
      const parsed = parseAtLocation(target)
      return parsed != null && lifetimeCount(values, `sold_${parsed.id}_at_${parsed.locationId}`) >= count
    }
    case 'kill_enemy':
      return target != null && lifetimeCount(values, `killed_${target}`) >= count
    case 'kill_enemy_class': {
      const parsed = parseKillEnemyClass(target)
      if (!parsed) return false
      return parsed.classes.some(
        (label) => lifetimeCount(values, `killed_${parsed.enemyId}_class_${label}`) >= count,
      )
    }
    case 'equip_set':
      return holdsEquipSet(save, target)
    case 'equip_quiver_bow': {
      if (!target) return false
      const worn = equippedItemIds(save)
      if (!worn.has(target)) return false
      const weaponId = save.equipment.slots[WEAPON_TOOL_SLOT_ID]?.itemId
      return weaponId != null && itemHasCapability(db, weaponId, 'bow_combat_xp')
    }
    case 'enchant':
      return lifetimeCount(values, 'items_enchanted') >= count
    case 'potion':
      return lifetimeCount(values, 'potions_created') >= count
    case 'spell_projects':
      return lifetimeCount(values, 'spell_projects') >= count
    case 'gold':
      return lifetimeCount(values, 'gold_earned') >= count
    case 'skill_all':
      if (typeof required !== 'number') return false
      return db.Skills.every((skill) => getSkillProgress(save, skill['Skill ID']).level >= required)
    default: {
      const skillId = achievement['Target Skill ID']
      if (!skillId || typeof required !== 'number') return false
      return getSkillProgress(save, skillId).level >= required
    }
  }
}

/** Refresh lifetime totals and unlock skill-level achievements. */
export function syncProgressionMeta(db: GameDatabase, save: PlayerSave, now = Date.now()): PlayerSave {
  const crittersCollected = (save.critterCollections ?? []).reduce(
    (sum, row) => sum + Math.max(0, row.count),
    0,
  )
  const values = {
    ...save.statistics.values,
    total_level: totalLevel(save),
    total_experience: totalSkillXp(save),
    gold_earned: Number(save.statistics.values.gold_earned ?? 0),
    monsters_killed: Number(save.statistics.values.monsters_killed ?? 0),
    critters_collected: crittersCollected,
    bounties_completed: Number(save.statistics.values.bounties_completed ?? 0),
  }

  let achievements = [...save.achievements]
  const unlockedAt = new Date(now).toISOString()
  for (const achievement of asAchievementRows(db)) {
    const achievementId = achievement['Achievement ID']
    if (achievement.Category === REVOCABLE_ACHIEVEMENT_CATEGORY) {
      achievements = holdsRevocableAchievement(save, achievementId)
        ? upsertAchievement(achievements, achievementId, unlockedAt)
        : revokeAchievement(achievements, achievementId)
      continue
    }
    if (holdsMilestone(db, save, achievement, values)) {
      achievements = upsertAchievement(achievements, achievementId, unlockedAt)
    }
  }

  return {
    ...save,
    statistics: { values },
    achievements,
  }
}

File: packages/ik_rules/lib/src/achievements/progress.dart
import 'dart:math' as math;

import 'package:collection/collection.dart';
import 'package:ik_content/ik_content.dart';

import '../activity/xp.dart';
import '../critters/critters.dart';
import '../equipment/loadout.dart';
import '../js_compat.dart';
import '../save/generated/save_models.dart';
import '../skills/totals.dart';
import '../tags.dart';
import '../time.dart';

/// The Achievements and Statistics tables stay untyped rows: nothing reads them
/// by column except this module and the panels that list them.
List<Map<String, Object?>> achievementRows(GameDatabase db) => db.achievements;

List<Map<String, Object?>> statisticRows(GameDatabase db) => db.statistics;

/// Achievements in this category are re-checked every sync and can be lost.
///
/// A skill milestone is a thing the player did once, so it is theirs forever. A
/// collection is a statement about the collection as it stands now, which stops
/// being true the moment the world grows a new critter.
const String revocableAchievementCategory = 'Collections';

const String critterCollectorAchievementId = 'ACH-0015';

const List<String> achievementDifficulties = <String>['Easy', 'Medium', 'Hard'];

PlayerSave addLifetimeStat(PlayerSave save, String key, [num amount = 1]) {
  final current = jsNumber(save.statistics.values[key] ?? 0);
  return save.copyWith(
    statistics: PlayerStatistics(
      values: <String, num>{...save.statistics.values, key: current + amount},
    ),
  );
}

bool isSpellProject(Map<String, Object?> project) {
  final key = jsString(project['Internal Key']);
  final name = jsString(project['Display Name']).toLowerCase();
  return key.contains('_spell') || (name.contains('spell') && !name.contains('enchant'));
}

PlayerSave recordProjectMilestones(GameDatabase db, PlayerSave save, String projectId, num crafts) {
  final project = db.projects.firstWhereOrNull(
    (row) => jsString(row.raw['Project ID']) == projectId,
  );
  var next = addLifetimeStat(save, 'project_$projectId', crafts);
  final locationId = save.currentLocationId;
  if (locationId.isNotEmpty) {
    next = addLifetimeStat(next, 'project_${projectId}_at_$locationId', crafts);
  }
  final outputId = jsString(project?.raw['Output Item / Target ID']);
  if (outputId.startsWith('ENCH-')) {
    next = addLifetimeStat(next, 'items_enchanted', crafts);
  }
  if (project != null && isSpellProject(project.raw)) {
    next = addLifetimeStat(next, 'spell_projects', crafts);
  }
  return next;
}

PlayerSave recordProductionMilestones(
  GameDatabase db,
  PlayerSave save,
  String outputItemId,
  num quantity,
) {
  var next = addLifetimeStat(save, 'output_$outputItemId', quantity);
  final locationId = save.currentLocationId;
  if (locationId.isNotEmpty) {
    next = addLifetimeStat(next, 'output_${outputItemId}_at_$locationId', quantity);
  }
  final item = db.items.firstWhereOrNull((row) => jsString(row.raw['Item ID']) == outputItemId);
  if (jsString(item?.raw['Category']) == 'Potion') {
    next = addLifetimeStat(next, 'potions_created', quantity);
  }
  return next;
}

PlayerSave recordFoodConsumed(PlayerSave save, String itemId) =>
    addLifetimeStat(save, 'consumed_$itemId');

PlayerSave recordGatheredDrops(
  PlayerSave save,
  Iterable<String> itemIds,
  String locationId,
  String? weaponId,
) {
  final wield = weaponId != null && weaponId.isNotEmpty ? weaponId : 'none';
  var next = save;
  for (final itemId in itemIds) {
    next = addLifetimeStat(next, 'gathered_${itemId}_at_${locationId}_wield_$wield');
  }
  return next;
}

PlayerSave recordItemsSoldAtLocation(
  PlayerSave save,
  Iterable<({String itemId, num quantity})> items,
  String locationId,
) {
  var next = save;
  for (final item in items) {
    final qty = math.max(0, jsNumber(item.quantity));
    if (qty <= 0) continue;
    next = addLifetimeStat(next, 'sold_${item.itemId}_at_$locationId', qty);
  }
  return next;
}

bool itemHasClassLabel(GameDatabase db, String itemId, String label) {
  final wanted = label.trim().toLowerCase();
  if (wanted.isEmpty) return false;
  if (itemHasCapability(db, itemId, wanted)) return true;
  final item = db.items.firstWhereOrNull((row) => jsString(row.raw['Item ID']) == itemId);
  return capabilityTags(item?.raw['Functional / Source Tags']).contains(wanted);
}

List<String> classLabelsFromAchievements(GameDatabase db) {
  final labels = <String>{};
  for (final row in achievementRows(db)) {
    if (jsString(row['Check Type']) != 'kill_enemy_class') continue;
    final parsed = _parseKillEnemyClass(
      row['Target ID'] is String ? row['Target ID'] as String : null,
    );
    if (parsed == null) continue;
    labels.addAll(parsed.classes);
  }
  return labels.toList();
}

List<String> classLabelsOnItem(GameDatabase db, String itemId) {
  return classLabelsFromAchievements(db)
      .where((label) => itemHasClassLabel(db, itemId, label))
      .toList();
}

PlayerSave recordEnemyKill(GameDatabase db, PlayerSave save, String enemyId) {
  var next = addLifetimeStat(save, 'killed_$enemyId');
  final weaponId = save.equipment.slots[weaponToolSlotId]?.itemId;
  if (weaponId == null) return next;
  for (final label in classLabelsOnItem(db, weaponId)) {
    next = addLifetimeStat(next, 'killed_${enemyId}_class_$label');
  }
  return next;
}

({String id, String locationId})? _parseAtLocation(String? target) {
  if (target == null) return null;
  final at = target.indexOf('@');
  if (at <= 0 || at == target.length - 1) return null;
  return (id: target.substring(0, at), locationId: target.substring(at + 1));
}

({String itemId, String locationId, String weaponId})? _parseGatherDrop(String? target) {
  if (target == null) return null;
  final parts = target.split('+wield:');
  if (parts.length != 2) return null;
  final parsed = _parseAtLocation(parts[0]);
  if (parsed == null || parts[1].isEmpty) return null;
  return (itemId: parsed.id, locationId: parsed.locationId, weaponId: parts[1]);
}

({String enemyId, List<String> classes})? _parseKillEnemyClass(String? target) {
  if (target == null) return null;
  final parts = target.split('+class:');
  if (parts.length != 2) return null;
  final classes = parts[1]
      .split('|')
      .map((part) => part.trim().toLowerCase())
      .where((part) => part.isNotEmpty)
      .toList();
  if (parts[0].isEmpty || classes.isEmpty) return null;
  return (enemyId: parts[0], classes: classes);
}

Set<String> _equippedItemIds(PlayerSave save) {
  return save.equipment.slots.values.map((stack) => stack?.itemId).whereType<String>().toSet();
}

bool _holdsEquipSet(PlayerSave save, String? target) {
  if (target == null || target.isEmpty) return false;
  final needed = target.split(',').map((part) => part.trim()).where((part) => part.isNotEmpty);
  if (needed.isEmpty) return false;
  final worn = _equippedItemIds(save);
  return needed.every(worn.contains);
}

List<AchievementProgress> _upsertAchievement(
  List<AchievementProgress> list,
  String achievementId,
  String unlockedAt,
) {
  final existing = list.firstWhereOrNull((row) => row.achievementId == achievementId);
  if (existing != null && existing.unlocked) return list;
  return <AchievementProgress>[
    ...list.where((row) => row.achievementId != achievementId),
    AchievementProgress(achievementId: achievementId, unlocked: true, unlockedAt: unlockedAt),
  ];
}

List<AchievementProgress> _revokeAchievement(List<AchievementProgress> list, String achievementId) {
  final existing = list.firstWhereOrNull((row) => row.achievementId == achievementId);
  if (existing == null) return list;
  return list.where((row) => row.achievementId != achievementId).toList();
}

/// Whether the collection holds at least one of every critter that exists.
bool hasEveryCritter(PlayerSave save) {
  if (critterDefs.isEmpty) return false;
  return critterDefs.every((critter) => collectionCount(save, critter.id) > 0);
}

/// Whether a save currently qualifies for a category that can be lost again.
bool _holdsRevocableAchievement(PlayerSave save, String achievementId) {
  return achievementId == critterCollectorAchievementId && hasEveryCritter(save);
}

num _lifetimeCount(Map<String, num> values, String key) => jsNumber(values[key] ?? 0);

bool _holdsMilestone(
  GameDatabase db,
  PlayerSave save,
  Map<String, Object?> achievement,
  Map<String, num> values,
) {
  final check = jsString(achievement['Check Type']);
  final target = achievement['Target ID'];
  final count = jsNumber(achievement['Required Count'] ?? 1);
  final required = achievement['Required Level'];
  switch (check) {
    case 'project':
      return target is String && _lifetimeCount(values, 'project_$target') >= count;
    case 'consume':
      return target is String && _lifetimeCount(values, 'consumed_$target') >= count;
    case 'output_item':
      return target is String && _lifetimeCount(values, 'output_$target') >= count;
    case 'output_at_location':
      final outputAt = _parseAtLocation(target is String ? target : null);
      return outputAt != null &&
          _lifetimeCount(values, 'output_${outputAt.id}_at_${outputAt.locationId}') >= count;
    case 'project_at_location':
      final projectAt = _parseAtLocation(target is String ? target : null);
      return projectAt != null &&
          _lifetimeCount(values, 'project_${projectAt.id}_at_${projectAt.locationId}') >= count;
    case 'gather_drop':
      final gathered = _parseGatherDrop(target is String ? target : null);
      return gathered != null &&
          _lifetimeCount(
                values,
                'gathered_${gathered.itemId}_at_${gathered.locationId}_wield_${gathered.weaponId}',
              ) >=
              count;
    case 'sold_at_location':
      final sold = _parseAtLocation(target is String ? target : null);
      return sold != null &&
          _lifetimeCount(values, 'sold_${sold.id}_at_${sold.locationId}') >= count;
    case 'kill_enemy':
      return target is String && _lifetimeCount(values, 'killed_$target') >= count;
    case 'kill_enemy_class':
      final killClass = _parseKillEnemyClass(target is String ? target : null);
      if (killClass == null) return false;
      return killClass.classes.any(
        (label) => _lifetimeCount(values, 'killed_${killClass.enemyId}_class_$label') >= count,
      );
    case 'equip_set':
      return _holdsEquipSet(save, target is String ? target : null);
    case 'equip_quiver_bow':
      if (target is! String || target.isEmpty) return false;
      if (!_equippedItemIds(save).contains(target)) return false;
      final weaponId = save.equipment.slots[weaponToolSlotId]?.itemId;
      return weaponId != null && itemHasCapability(db, weaponId, 'bow_combat_xp');
    case 'enchant':
      return _lifetimeCount(values, 'items_enchanted') >= count;
    case 'potion':
      return _lifetimeCount(values, 'potions_created') >= count;
    case 'spell_projects':
      return _lifetimeCount(values, 'spell_projects') >= count;
    case 'gold':
      return _lifetimeCount(values, 'gold_earned') >= count;
    case 'skill_all':
      if (required is! num) return false;
      return db.skills.every((skill) => getSkillProgress(save, skill.skillId).level >= required);
    default:
      final skillId = achievement['Target Skill ID'];
      if (skillId is! String || skillId.isEmpty || required is! num) return false;
      return getSkillProgress(save, skillId).level >= required;
  }
}

/// Refreshes the lifetime totals and unlocks any skill-level achievement the
/// player now qualifies for.
PlayerSave syncProgressionMeta(GameDatabase db, PlayerSave save, num nowMs) {
  final crittersCollected = save.critterCollections.fold<num>(
    0,
    (sum, row) => sum + math.max(0, row.count),
  );
  final values = <String, num>{
    ...save.statistics.values,
    'total_level': totalLevel(save),
    'total_experience': totalSkillXp(save),
    'gold_earned': jsNumber(save.statistics.values['gold_earned'] ?? 0),
    'monsters_killed': jsNumber(save.statistics.values['monsters_killed'] ?? 0),
    'critters_collected': crittersCollected,
    'bounties_completed': jsNumber(save.statistics.values['bounties_completed'] ?? 0),
  };

  var achievements = save.achievements;
  final unlockedAt = isoFromMs(nowMs);
  for (final achievement in achievementRows(db)) {
    final achievementId = jsString(achievement['Achievement ID']);
    if (achievement['Category'] == revocableAchievementCategory) {
      achievements = _holdsRevocableAchievement(save, achievementId)
          ? _upsertAchievement(achievements, achievementId, unlockedAt)
          : _revokeAchievement(achievements, achievementId);
      continue;
    }
    if (_holdsMilestone(db, save, achievement, values)) {
      achievements = _upsertAchievement(achievements, achievementId, unlockedAt);
    }
  }

  return save.copyWith(
    statistics: PlayerStatistics(values: values),
    achievements: achievements,
  );
}
